### PR TITLE
Adding -stop-after lambda flag option in ocaml{c,opt}

### PR DESCRIPTION
**Pull request by dsmith47**
Part of #6102, adding logic to stop compilation after lambdas are generated but before code is written.

Allows outputting of just lambda with ocaml compilers:
```
ocaml -dlambda -stop-after lambda script.ml

ocamlopt -dlambda -stop-after lambda script.ml
```

Required branches checking if compilation should be executed in driver/ocamlc.ml and driver/ocamlopt.ml

### CODE REVIEW

File: driver/optcompile.ml [hunk 1]
```diff
@@ -31,6 +31,21 @@ let (|>>) (x, y) f = (x, f y)
 
 (** Native compilation backend for .ml files. *)
 
+let flambda_bytecode i backend required_globals ((module_ident, size), lam) =
+  ((module_ident, size), lam)
+  |> (fun ((module_ident, size), lam) ->
```
Comment (gasche) [inline driver/optcompile.ml:36]: this looks like it could be simplified away

**Review by gasche — COMMENTED** [review]
Thanks for your contributing! Implementing this feature unfortunately requires non-trivial restructuring of the driver/ files, and it seems you did a very good first attack on this problem.

The changes to compile.{ml,mli} look mostly fine to me. I wonder if we should maybe rename the `to_<pass>` functions to indicate also the starting point: `typedtree_to_lambda`, `lambda_to_bytecode`; for this the opinion of other `driver/` hackers (@Drup, @nojb ?) would be helpful.

I haven't looked at the changes in `optcompile.{ml,mli}` in details yet, but I'm not convinced by the occurrence of the name `bytecode` there, given that the bytecode is never used as an intermediate representation in the native-compilation pipeline (which goes like typedtree -> lambda -> clambda -> cmm -> ..., if I remember correctly). This naming needs to change -- maybe you have a suggestion right now, or maybe one will emerge after I or someone else takes a closer look.

**Comment by gretay-js** [comment]
Coincidentally,  I am working on a patch to save different IRs to files instead of printing them out as we currently do with -dcmm etc.  It is based on @mshinwell's code here:
https://github.com/mshinwell/ocaml/commits/save_ir

**Comment by gasche** [comment]
@gretay-js ; Thanks for the heads-up. This also sounds related to the discussion we had in #1913 which introduced the `-dump-into-file` option.

**Comment by gasche** [comment]
@dsmith47 gentle ping? What's the status of this PR for you? Are you planning to work more to polish this PR, or would you like other people to to consider finishing the work?

**Comment by nojb** [comment]
This is superseded by #11863

**Pull request closed without merging**